File: app/src/main/java/com/aspirasoft/huntrek/view/AnnotatedProgressBar.java
package com.aspirasoft.huntrek.view;

import android.content.Context;
import android.os.Build;
import android.os.Handler;
import android.util.AttributeSet;
import android.widget.ProgressBar;
import androidx.annotation.RequiresApi;

import java.util.Random;

/**
 * Created by saifkhichi96 on 24/12/2017.
 */

public class AnnotatedProgressBar extends ProgressBar {
    private final Handler mLoadHandler = new Handler();
    private final Random mRandomGenerator = new Random();
    private boolean finishing = false;

    public AnnotatedProgressBar(Context context) {
        super(context);
    }

    public AnnotatedProgressBar(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    public AnnotatedProgressBar(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }

    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
    public AnnotatedProgressBar(Context context, AttributeSet attrs, int defStyleAttr, int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);
    }

    void start(Context context) {
        setProgress(0);

        int initialDelay = 250 + mRandomGenerator.nextInt(250);
        mLoadHandler.postDelayed(new ProgressUpdater(), initialDelay);
    }

    void finish() {
        finishing = true;
    }

    boolean isFinished() {
        return getProgress() == getMax();
    }

    private class ProgressUpdater extends Thread {
        @Override
        public void run() {
            // Advance one step at every iteration. Cap at 90% progress.
            int currentProgress = getProgress();
            if (!finishing && currentProgress < 0.90 * getMax()) {
                int newProgress = (int) (currentProgress + 0.01 * getMax());
                setProgress(newProgress);

                // Schedule next iteration
                int delay = 250 + mRandomGenerator.nextInt(250);
                mLoadHandler.postDelayed(this, delay);
            } else if (finishing && !isFinished()) {
                int newProgress = (int) (currentProgress + 0.01 * getMax());
                setProgress(newProgress);

                // Schedule next iteration
                int delay = 25;
                mLoadHandler.postDelayed(this, delay);
            }
        }
    }

}

File: app/src/main/java/dev/aspirasoft/huntrek/ui/view/AnnotatedProgressBar.kt
package dev.aspirasoft.huntrek.ui.view

import android.content.Context
import android.os.Handler
import android.util.AttributeSet
import android.widget.ProgressBar
import java.util.*

/**
 * Created by saifkhichi96 on 24/12/2017.
 */
class AnnotatedProgressBar(context: Context, attrs: AttributeSet?, defStyleAttr: Int) :
    ProgressBar(context, attrs, defStyleAttr) {

    constructor(context: Context, attrs: AttributeSet?) : this(context, attrs, 0)
    constructor(context: Context) : this(context, null)

    private val mLoadHandler = Handler()
    private val mRandomGenerator = Random()

    private var finishing = false

    fun start(context: Context?) {
        progress = 0
        val initialDelay = 250 + mRandomGenerator.nextInt(250)
        mLoadHandler.postDelayed(ProgressUpdater(), initialDelay.toLong())
    }

    fun finish() {
        finishing = true
    }

    val isFinished: Boolean
        get() = progress == max

    private inner class ProgressUpdater : Thread() {
        override fun run() {
            // Advance one step at every iteration. Cap at 90% progress.
            val currentProgress = progress
            if (!finishing && currentProgress < 0.90 * max) {
                val newProgress = (currentProgress + 0.01 * max).toInt()
                progress = newProgress

                // Schedule next iteration
                val delay = 250 + mRandomGenerator.nextInt(250)
                mLoadHandler.postDelayed(this, delay.toLong())
            } else if (finishing && !isFinished) {
                val newProgress = (currentProgress + 0.01 * max).toInt()
                progress = newProgress

                // Schedule next iteration
                val delay = 25
                mLoadHandler.postDelayed(this, delay.toLong())
            }
        }
    }

}